sdk/python/lib/pulumi/provider/server.py

```diff
@@ -112,19 +113,28 @@ async def Construct(
         await self.lock.acquire()
         try:
             return await self._construct(request, context)
-        except InputPropertiesError as e:
-            status = self.create_grpc_invalid_properties_status(e.message, e.errors)
-            await context.abort_with_status(status)
-            # We already aborted at this point
-            raise
-        except InputPropertyError as e:
-            status = self.create_grpc_invalid_properties_status(
-                "", [{"property_path": e.property_path, "reason": e.reason}]
-            )
-            await context.abort_with_status(status)
-            # We already aborted at this point
-            raise
-
+        except Exception as e:  # noqa
+            if isinstance(e, InputPropertiesError):
+                status = self.create_grpc_invalid_properties_status(e.message, e.errors)
+                await context.abort_with_status(status)
+                # We already aborted at this point
+                raise
+            elif isinstance(e, InputPropertyError):
+                status = self.create_grpc_invalid_properties_status(
+                    "", [{"property_path": e.property_path, "reason": e.reason}]
+                )
+                await context.abort_with_status(status)
+                # We already aborted at this point
+                raise
+            else:
+                stack = traceback.extract_tb(e.__traceback__)[:]
+                if len(stack) > 3:
+                    # If we have more than 3 frames, we've made it to the
+                    # provider code: server.Construct > server._construct >
+                    # provider.construct. Drop our internal frames.
+                    stack = stack[3:]
```

Comment:
Will this give a reasonable trace if there is an error in our SDK code?

Reply:
It mostly does. Thinking more about this, we correctly handle the error on the way into the user's code, but if we fail on the way out, for example during serialization, we might truncate part of our stack.

I think I can make this behave better by adding a handler with a custom exception around the SDK <> component boundary.